### ISSUE TITLE
`matvec` inconsistent behavior when used with the `col` argument.

### ISSUE THREAD

**Issue by MartinStancsicsQC** [Quantco/tabmat#256]
## Description

`matvec` has inconsistent behavior between the various matrix types, and also compared to the docs. According to the docs, `X.matvec(b, cols=cols)` returns the product `X[:, cols] @ b`. In reality, this is not always the case, and the behavior differs between various matrix types.

## Example

```python
import numpy as np
import tabmat as tm

X_dense = tm.DenseMatrix(np.array([
    [1., 2., 3.],
    [4., 5., 6.],
]))
X_split = tm.SplitMatrix([X_dense])
```

Consider the following cases (`X` stands for both `X_dense` and `X_split`):

- `X.matvec([1., 1., 1.])`
  - Expected: `[6, 15]`
  - Got: `[6, 15]` (`DenseMatrix.matvec` returns a `DenseMatrix` instead of a `numpy.ndarray`, but that's probably fine as that is a subclass)

- `X.matvec([1., 1., 1.], cols=[0, 1])`
  - Expected: Error because of shape mismatch
  - Got: `[3, 9]` in both cases
  - Note: It is not necessarily a bug, as `X[:, cols] @ b[cols]` makes a lot of sense, but still inconsistent with the docs

- `X.matvec([1., 1.], cols=[0, 1])`
  - Expected: `[3, 9]`
  - Dense: `[3, 9]`
  - Split: `ValueError: shapes (2, 3) and (2,) not aligned`

- `X.matvec([1.], cols=[0, 1])``
  - Expected: Error because of shape mismatch
  - Dense: `[1, 4]`
  - Split: `ValueError: shapes (2, 3) and (1,) not aligned`

## Environment

Using `tabmat=3.1.7`, conda-forge `osx-arm64` build.

<details>

```
# Name                    Version                   Build  Channel
brotli                    1.0.9                h1a8c8d9_8    conda-forge
brotli-bin                1.0.9                h1a8c8d9_8    conda-forge
bzip2                     1.0.8                h3422bc3_4    conda-forge
ca-certificates           2023.5.7             hf0a4a13_0    conda-forge
certifi                   2023.5.7           pyhd8ed1ab_0    conda-forge
charset-normalizer        3.1.0              pyhd8ed1ab_0    conda-forge
idna                      3.4                pyhd8ed1ab_0    conda-forge
libblas                   3.9.0           17_osxarm64_openblas    conda-forge
libbrotlicommon           1.0.9                h1a8c8d9_8    conda-forge
libbrotlidec              1.0.9                h1a8c8d9_8    conda-forge
libbrotlienc              1.0.9                h1a8c8d9_8    conda-forge
libcblas                  3.9.0           17_osxarm64_openblas    conda-forge
libcxx                    16.0.5               h4653b0c_0    conda-forge
libffi                    3.4.2                h3422bc3_5    conda-forge
libgfortran               5.0.0           12_2_0_hd922786_31    conda-forge
libgfortran5              12.2.0              h0eea778_31    conda-forge
libjemalloc-local         5.3.0                hb7217d7_0    conda-forge
liblapack                 3.9.0           17_osxarm64_openblas    conda-forge
libopenblas               0.3.23          openmp_hc731615_0    conda-forge
libsqlite                 3.42.0               hb31c410_0    conda-forge
libzlib                   1.2.13               h03a7124_4    conda-forge
llvm-openmp               16.0.5               h1c12783_0    conda-forge
ncurses                   6.3                  h07bb92c_1    conda-forge
numpy                     1.24.3          py310haa1e00c_0    conda-forge
openssl                   3.1.1                h53f4e23_1    conda-forge
packaging                 23.1               pyhd8ed1ab_0    conda-forge
pandas                    2.0.2           py310h1cdf563_0    conda-forge
pip                       23.1.2             pyhd8ed1ab_0    conda-forge
platformdirs              3.5.1              pyhd8ed1ab_0    conda-forge
pooch                     1.7.0              pyha770c72_3    conda-forge
pysocks                   1.7.1              pyha2e5f31_6    conda-forge
python                    3.10.11         h3ba56d0_0_cpython    conda-forge
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python-tzdata             2023.3             pyhd8ed1ab_0    conda-forge
python_abi                3.10                    3_cp310    conda-forge
pytz                      2023.3             pyhd8ed1ab_0    conda-forge
readline                  8.2                  h92ec313_1    conda-forge
requests                  2.31.0             pyhd8ed1ab_0    conda-forge
scipy                     1.10.1          py310h0975f3d_3    conda-forge
setuptools                67.7.2             pyhd8ed1ab_0    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
tabmat                    3.1.7           py310h07e7c9c_0    conda-forge
tk                        8.6.12               he1e0b03_0    conda-forge
typing-extensions         4.6.3                hd8ed1ab_0    conda-forge
typing_extensions         4.6.3              pyha770c72_0    conda-forge
tzdata                    2023c                h71feb2d_0    conda-forge
urllib3                   2.0.2              pyhd8ed1ab_0    conda-forge
wheel                     0.40.0             pyhd8ed1ab_0    conda-forge
xz                        5.2.6                h57fd34a_0    conda-forge
```

</details>

**Comment by MarcAntoineSchmidtQC** [comment]
Thanks for opening the issue. `matvec` is inconsistent between types only when it is being "incorrectly used". The docs is actually wrong. `matvec` is doing `X[:, cols] @ b[cols]`. Therefore, `b` needs to be of the correct dimension for the result to be valid. if X is 2x3, then the vector needs to of length 3.

So, the issues this raises are:
1. The documentation is wrong or misleading.
2. Tabmat suffers from a lack of input validation. We should clearly add this.

**Comment by MartinStancsicsQC** [comment]
Thanks for the clarification! Hopefully #257 addresses the issues you've mentioned.

**Comment by MarcAntoineSchmidtQC** [comment]
Fixed by #257
